### PR TITLE
feat(client): Add `omit` option to the queries

### LINKED ISSUE #5042: Add way to exclude fields via query arguments or globally
## Problem

Sometimes it is helpful to just remove some fields from the selection set like `password` so that we accidentally don't expose them to the frontend.

## Proposals

Adding a `exclude` key in the arguments of the queries to easily exclude fields from the selectionSet just like `include` or `select`.

```js
prisma.user.findMany({
  exclude:{
    password: true
  }
})
```

Or

```js
await prisma.user.findMany({
  omit: {
    password: true,
  },
})
```

## Alternatives

1. Use query and list all the fields you want
1. Schema annotation
2. Manual omission via lodash or something !?

## Related
- https://github.com/prisma/prisma/issues/2170
- https://github.com/prisma/prisma/issues/3246

### PR DESCRIPTION
**Pull request by SevInf**
Counterpart to prisma/prisma-engines#4807

Can be used standalone or combined with `include`, allows to exclude
the fields that normally would be included by default. Available in all
methods that return actual database records.

It is not useable together with `select` and attempt to do so would
cause type check and validation error.

When using together with result extensions, excluded dependency of a
computed field will be queried from a DB, but will not be returned to
the end user, unless computed field is exlucded as well (see "omit
with extensions" tests in this PR). This behaviour is equivalent to what
we do if depenency of a computed field is not mentioned in explicit
`select`.

TODO:
- [x] preview feature
- [x] validation of non-existing fields in exclude

Close prisma/team-orm#1080
#5042

/integration

### CODE REVIEW

File: packages/client/package.json [hunk 1]
```diff
@@ -174,7 +174,7 @@
     "@prisma/debug": "workspace:*",
     "@prisma/driver-adapter-utils": "workspace:*",
     "@prisma/engines": "workspace:*",
-    "@prisma/engines-version": "5.13.0-6.dcdb692a8946281c0c85b4a6f7081984eec92b92",
+    "@prisma/engines-version": "5.13.0-3.feat-exclude-selection-e76495825e7475c1f31801a6b5e233caa6edebb8",
```
Comment (SevInf) [inline packages/client/package.json:177]: TODO: revert before merging

File: packages/client/src/generation/TSClient/Args.ts [hunk 2]
```diff
@@ -1,144 +1,118 @@
...
+  getLegacyModelArgName,
+  getModelArgName,
+  getSelectName,
+} from '../utils'
 import { GenerateContext } from './GenerateContext'
 import { getArgFieldJSDoc } from './helpers'
-import { InputField } from './Input'
+import { buildInputField } from './Input'
+
+export class ArgsTypeBuilder {
```
Comment (SevInf) [inline packages/client/src/generation/TSClient/Args.ts:15]: This replaces both `ArgsType` and `MinArgsType` type classes - we now can selectively decide how to build different args types and which fields to include.

File: packages/client/src/runtime/core/errorRendering/applyValidationError.test.ts [hunk 3]
```diff
@@ -622,8 +776,6 @@ describe('UnknownSelectionField', () => {
         include: {
           notThere: true,
           ~~~~~~~~
-      ?   id?: true,
```
Comment (SevInf) [inline packages/client/src/runtime/core/errorRendering/applyValidationError.test.ts:625]: Opprtunistic fix: we previously suggested to use scalar fields in `include`. Only relations are valid in include, so those suggestions were incorrect.

**Comment by SevInf** [comment]
Note: tests are failing because https://github.com/prisma/prisma/pull/23776 needs integration first, it should not block the review though.

File: packages/client/src/runtime/core/types/ValidationError.ts [hunk 4]
```diff
@@ -9,10 +9,12 @@ import { EngineValidationError, OutputTypeDescription } from '../engines'
...
+ * Pair of mutually exclusive fields are found on selection (for example select + include or select + exclude)
  */
-export type IncludeAndSelectError = {
-  kind: 'IncludeAndSelect'
+export type MutuallyExclusiveFieldsError = {
+  kind: 'MutuallyExclusiveFields'
+  firstField: string
+  secondField: string
   selectionPath: string[]
 }
```
Comment (Druue) [inline packages/client/src/runtime/core/types/ValidationError.ts:19]: ✨

File: packages/client/src/runtime/core/jsonProtocol/serializeJsonQuery.ts [hunk 5]
```diff
@@ -90,29 +101,50 @@ function serializeFieldSelection(
   { select, include, ...args }: JsArgs = {},
   context: SerializeContext,
 ): JsonFieldSelection {
+  let exclude: Exclusion | undefined
+  if (context.isPreviewFeatureOn('excludeApi')) {
+    exclude = args.exclude
+    delete args.exclude
```
Comment (Druue) [inline packages/client/src/runtime/core/jsonProtocol/serializeJsonQuery.ts:107]: TIL about the `delete` keyword, pretty cool :o

File: packages/client/src/runtime/core/types/exported/Result.ts [hunk 6]
```diff
@@ -48,7 +48,16 @@ export type FluentOperation =
...
+export type TrueKeys<T> = {
+  [K in keyof T]: T[K] extends true ? K : never
+}[keyof T]
+
+export type GetFindResult<P extends OperationPayload, A> = A extends { exclude: infer Exclusion }
+  ? Compute<Omit<GetSelectIncludeResult<P, A>, TrueKeys<Exclusion>>>
+  : GetSelectIncludeResult<P, A>
+
+// prettier-ignore
+export type GetSelectIncludeResult<P extends OperationPayload, A> =
```
Comment (Druue) [inline packages/client/src/runtime/core/types/exported/Result.ts:60]: 🤯

Reading through the [infer docs](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types) but I really appreciate the explanation of:
```ts
type ElementType<A> = A extends Array<infer X>
    ? x
    : never
```
translating to 
```rs
match A {
    Array(x) => x,
    _ => never,
}
```

File: packages/client/src/runtime/core/errorRendering/applyValidationError.ts [hunk 7]
```diff
@@ -439,6 +486,22 @@ function addSelectionSuggestions(selection: ObjectValue, outputType: OutputTypeD
...
+}
+
+function addExclusionSuggestions(selection: ObjectValue, outputType: OutputTypeDescription) {
+  for (const field of outputType.fields) {
+    if (!selection.hasField(field.name) && !field.isRelation) {
+      selection.addSuggestion(new ObjectFieldSuggestion(field.name, 'true'))
+    }
+  }
+}
+
```
Comment (Druue) [inline packages/client/src/runtime/core/errorRendering/applyValidationError.ts:504]: Are these for the typescript language-server or am I misunderstanding what `suggestion` means

File: packages/client/src/generation/ts-builders/TypeBuilder.ts [hunk 8]
```diff
@@ -3,7 +3,10 @@ import type { KeyType } from './KeyType'
 import { Writer } from './Writer'
 
 export abstract class TypeBuilder implements BasicBuilder {
+  // TODO: this should be replaced with precedence system that would
+  // automatically add parenthesis where they are needed
```
Comment (Druue) [inline packages/client/src/generation/ts-builders/TypeBuilder.ts:7]:
```suggestion
  // TODO(@SevInf): this should be replaced with precedence system that would
  // automatically add parenthesis where they are needed
```

**Review by Druue — COMMENTED** [review]
🗒️ to be continued... 👀

File: packages/client/src/runtime/core/extensions/resultUtils.ts [hunk 9]
```diff
@@ -122,3 +122,24 @@ export function applyComputedFieldsToSelection(
   }
   return result
 }
+
+export function applyComputedFieldsToExclusion(
```
Comment (Druue) [inline packages/client/src/runtime/core/extensions/resultUtils.ts:126]: Add doc comment and explain what it does
MAybe rename to `getEngineExclusion`

File: packages/client/src/runtime/core/extensions/resultUtils.ts [hunk 10]
```diff
@@ -122,3 +122,24 @@ export function applyComputedFieldsToSelection(
...
+  exclusion: Exclusion,
+  computedFields: ComputedFieldsMap | undefined,
+): Selection {
```
Comment (Druue) [inline packages/client/src/runtime/core/extensions/resultUtils.ts:129]:
```suggestion
): Exclusion {
```

**Review by Druue — APPROVED** [review]

File: packages/client/package.json [hunk 11]
```diff
@@ -174,7 +174,7 @@
     "@prisma/debug": "workspace:*",
     "@prisma/driver-adapter-utils": "workspace:*",
     "@prisma/engines": "workspace:*",
-    "@prisma/engines-version": "5.13.0-16.6d35870e9578a3e7406168f61f89f84e78c1b38d",
+    "@prisma/engines-version": "5.13.0-20.feat-exclude-selection-8eed76cebeb3eb147fdfbf46e79211a543caa112",
```
Comment (SevInf) [inline packages/client/package.json:177]: TODO: remvoe before merging

**Pull request merged**